### ISSUE TITLE
[QUESTION] How to get mocha context and currentTest object

### ISSUE THREAD

**Issue by viraxslot** [testdeck/testdeck#279]
## Question
Hi, guys!
I see [this](https://github.com/testdeck/testdeck/issues/17) and [that](https://github.com/testdeck/testdeck/issues/54) tickets, but I still don't understand how to get mocha context using your approach with decorators.

When I'm trying to import `context` from @testdeck/mocha I got:
![image](https://user-images.githubusercontent.com/20737672/105493196-1b885200-5cca-11eb-8c09-fc6e0f16cdba.png)

If you could give an example how to get the context in mocha [root hooks](https://mochajs.org/#root-hook-plugins) it would be really great. Thanks for your help!

Btw I use "@testdeck/mocha": "^0.1.2"

**Comment by silkentrance** [comment]
https://github.com/testdeck/testdeck/tree/mocha-typescript/v1.1.17

@pana-cc implemented this in the above version.

I think that this change was lost during the rewrite and the added support for multiple test frameworks.

**Comment by viraxslot** [comment]
@silkentrance hi, thanks for the answer. Am I right `mocha-typescript` is deprecated? Any possibility to see this change in `@testdesk/mocha` package?

**Comment by silkentrance** [comment]
@viraxslot yes, mocha-typescript was deprecated in favor of `@testdeck/mocha`.  as for porting this: i need to figure out how to do this without breaking the interface of the jest/jasmine stuff, but i think that it is doable. it will take some time though, as i am currently preparing for my final exams...

**Comment by viraxslot** [comment]
@silkentrance thanks a lot, could you please tag me when it'll fixed? break a leg! :)

**Comment by KsenyLa** [comment]
Hi @silkentrance!
I'd really like to see that feature implemented, it would be awesome if you do it! Access to mochaContext is really important if you use mochawesome with added screenshots and logs to the report.

**Comment by ThatOhio** [comment]
Really wish I had this feature right now! +1

**Comment by silkentrance** [comment]
@ThatOhio @KsenyLa ATM i am preparing for my exams, so I do not have the time.
As for root hooks, this can still be done using the standard mechanisms, as these are not under the control of testdeck/mocha.

**Comment by viraxslot** [comment]
@silkentrance hey hey, sorry for pushing the issue, any updates about it?

**Comment by pana-cc** [comment]
Maybe I will take a look, I don't promise anything.

**Comment by pana-cc** [comment]
Ok, in the mocha-typescript the context was exposed in something like:
```
@suite class Hello {
  @context mochaContext: MochaContext;
  @test world() {
    this.mochaContext; // <- here
  }
}
```

It had some issues however with inheritance - what if multiple classes define @context with different fields? The previous implementation I think was assigning the field of the name somewhere on the test class but couldn't define multiple.

Then typing was required every time it is about to be used. On the other hand I remember that Angular wanted to expose some of its lifecycle hooks with symbols like:
```
import onInit from "@angular/angular";
class Component {
  [onInit]() {}
}
```
But as far as I can see they still don't. TypeScript was supposed to provide typings for what onInit can be used with.

In our case in similar way can be provided the context like:
```
import { context } from "@testdeck/mocha";

@suite Hello {
  static before() {
    this[context].slow(15);
  }
  before() {
    this[context].slow(30);
  }
  @test world() {
    this[context].slow(20):
  }
}
```

And by extension it also makes perfect sense for the `before` and `after` lifecycle hooks for the testdeck to be symbols.

Implementation is trivial - we can either wrap all functions that are passed to `wrap` with additional call that sets and clears the context, or do it everywhere `wrap` is called.

There is a branch now with an inital commit: https://github.com/testdeck/testdeck/commit/0eb6f795adc91c833f92e0696c5507ebf745265f#

**Comment by pana-cc** [comment]
And here is the TypeScript issue describing typings for symbols:
https://github.com/microsoft/TypeScript/issues/5579
Seems like there is a PR merged.

**Comment by pana-cc** [comment]
The `unique symbol` types: https://devblogs.microsoft.com/typescript/announcing-typescript-2-7/#unique-symbol-types-and-const-named-properties

**Comment by pana-cc** [comment]
The typings go like:
```
interface MochaContext {
    slow(ms: number): void;
    currentTest: object;
}

declare const context: unique symbol;

interface Object {
    [context]: MochaContext;
}

class Hello {
    world(): void {
        this[context].slow(10);
    }
}
```

**Comment by silkentrance** [comment]
@pana-cc while the solution would work, i find that it is kind of redundant to have the context since one can always use the decorators on the suite or on individual test methods to have slow() and so on.

and access to the current test is something that would make sense in a global hook that would then display information on the current test, so I also see no benefit of being able to access it from inside a test suite or test even.

which is why I pointed out that one can always use the root hooks:

>>> As for root hooks, this can still be done using the standard mechanisms, as these are not under the control of testdeck/mocha.

which will then use the standard mocha mechanisms.

**Comment by pana-cc** [comment]
We are wrapper for the base test frameworks, it makes sense to expose their context, in cases there is functionality that is not exposed through decorators, or in case that would make easier using existing libraries that are designed for the test framework instead of testdeck.

**Comment by silkentrance** [comment]
Here is the official Mocha context API/Impl. https://mochajs.org/api/context.js.html
where the runnable is https://mochajs.org/api/runnable.js.html.

Here is the mocha awesome documentation, here, addContext is the part that the users are really interested in: https://www.npmjs.com/package/mochawesome

Notice that it can be called from also inside the before/after hooks, for which we currently do not have a mocha context for,
here the context must be passed in as well, e.g.

- beforeAll: https://github.com/testdeck/testdeck/blob/c6e1afcf61d5b474529c2754755730a26b89167a/packages/core/index.ts#L91
- beforeEach: https://github.com/testdeck/testdeck/blob/c6e1afcf61d5b474529c2754755730a26b89167a/packages/core/index.ts#L114
- afterEach: https://github.com/testdeck/testdeck/blob/c6e1afcf61d5b474529c2754755730a26b89167a/packages/core/index.ts#L204
- afterAll: https://github.com/testdeck/testdeck/blob/c6e1afcf61d5b474529c2754755730a26b89167a/packages/core/index.ts#L222

which makes it even more complicated.

**Comment by silkentrance** [comment]
Long story short, addContext expects the Runnable interface, which is the mocha context object that can be accessed from inside the suite/test callbacks and the typings for that interface should match the runnable/test or suite, depending on the context (before/afterAll, before/afterEach, test, suite constructor) etc.

**Comment by pana-cc** [comment]
No no, it doesn't make it more complicated, this is basically everywhere we "wrap" the functions we create with pretty names, if you check my commit you will find it will handle before/after each/all. It could even be done by replacing the test function in "wrap" but it makes deep callstack, and node had limit on the callstack in some cases.

**Comment by pana-cc** [comment]
And about the tapings you are right, I don't have a solution yet to provide good typing for test or suite functions. May just leave it "any".

**Comment by silkentrance** [comment]
Ah, okay, haven't seen that. I also see that beforeAll/afterAll is not required as they do not have a context object or runnable.
And I also see that there is no context for the suite constructor either, so the runnable should eventuall be all instances of the mocha test class, or shouldn't it?

**Comment by silkentrance** [comment]
what about defining the interface like so

```
interface TestContext {}

interface MochaContext extends TestContext {
   readonly title: string;
   fullTitle(): string;
   titlePath(): string;
   currentRetry(): number;
   isPassed(): boolean;
   isFailed(): boolean;
   isPending(): boolean;
   ...
}
```

then there are the combined setter/getter like ```retries(), slow(), ...```.



the other implementations would then have an empty context object, e.g.

```
interface class JestTestContext extends TestContext {}
```

of course, having these empty base interface is somewhat frowned upon...

**Comment by pana-cc** [comment]
I am planning to set the context to this everywhere, even in beforeAll/afterAll, and it will hold the `this` passed to them. I am not sure if mocha provided meaningful context to before. Is it possible to set timeout in before? IDK I should check it.

The constructor is a good catch, we cant set the context to `this` obviously, maybe we can set it on the static context on the class.

```
@suite Hello {
  constructor() {
    Hello[context].timeout();
  }
}
```

It will be a mess if runs in parallel. :( better skip this for the constructor.

About the Mocha and Jest, and the rest, the "context" will be defined properly in their typings and in the core will be just "any".

like 
```
module "@testdeck/mocha" {
  export const context: unique symbol;
  interface Object {
    [context]: MochaContext;
  }
}
module "@testdeck/jest" {
  interface Object {
    [context]: JestContext;
  }
}
```
And when you import @testdeck/jest's context, this[context] will be jest's context. When you import @testdeck/mocha's context, this[context] will be mocha's context.

**Comment by pana-cc** [comment]
I just added a commit to set the “context” on the proper instances. It now needs:
 - tests
 - typings
 - documentation
 - major version bump? TBD...

This is now implemented a lot simple than it used to be, but also introduces a slight breaking change.

**Comment by silkentrance** [comment]
@pana-cc can you please make this a PR so that we can integrate this into learna-changelog?

**Comment by silkentrance** [comment]
NB. learna-changelog is PR based and gets its information from github...

**Comment by Kupstas** [comment]
Any news?

**Comment by pana-cc** [comment]
@Kupstas Yes, I've just pushed another commit with tests for the core. It will get done eventually. Still needs some tests for the mocha module itself and maybe jest and the rest.

There is a plot hole I've found that scares me at this point, for every class test method a before-each/test/after-each pair is executed where in before-each an instance of the class is created and captured in closure shared with the test and after-each. But that would fail terribly for parallel tests - async tests that for example send http requests and would try to create several instances at the same time, run the test methods concurrently, etc. I think instead of relying on the closure, the test instance that is created can be attached to the test context for mocha, but I am not sure for the rest... It is getting late today and I am definitely not going to resolve that problem in this PR, but may do it in future.

Also I've just found VSCode is working on APIs for test exploration and execution. We already have a lot of the functionality in the tests, don't be surprised if one day testdeck gets a runner on its own that comes with extension for VSCode test exploration, execution and debugging.

The promised PR: https://github.com/testdeck/testdeck/pulls

**Comment by pana-cc** [comment]
I just added a commit with the typing for mocha. Was a bit tricky. What is left is:
 - [ ] copy the typing for jest and jasmine
 - [ ] update docs
 - [ ] figure out versioning

**Comment by panayot-cankov** [comment]
jest and jasmine have no meaningful context (this) for the callback functions

**Comment by pana-cc** [comment]
It is now merged, officially you will access the mocha context using something like:
```
@suite class MyClass {
  @test method() {
    this[context] // <- This field will be set by the testdeck framework to the mocha context
  }
}
```

**Comment by pana-cc** [comment]
Still need docs.

**Comment by KsenyLa** [comment]
It has been almost one year since issue was created and we still hope for the solution. Could you publish it please?

**Comment by pana-cc** [comment]
@KsenyLa it should be available in v0.2.0 I've just publish. Give it a go, and please let me know if there are issues with it.

**Comment by KsenyLa** [comment]
@pana-cc, thank you so much for implementing this! It's a very important feature, and seems working fine for me
